File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
def solution(n, wires):
    answer = n+1
    arr = [[] for _ in range(n+1)]
    
    def bfs(start):
        q = deque([start])
        visited = [0]*(n+1)
        visited[start] = 1
        cnt = 1
        while q:
            s = q.popleft()
            
            for i in arr[s]:
                if visited[i] != 1:
                    visited[i] = 1
                    q.append(i)
                    cnt +=1
        return cnt
        
    
    for a,b in wires:
        arr[a].append(b)
        arr[b].append(a)
    
    for a,b in wires:
        arr[a].remove(b)
        arr[b].remove(a)
        
        answer = min(abs(bfs(a) - bfs(b)),answer)
        
        arr[a].append(b)
        arr[b].append(a)
    
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    static List<Integer>[] graph; 
    static int min;
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        graph = new ArrayList[n+1];
        
        for(int i=0;i<=n;i++){
            graph[i] = new ArrayList<>();
        }
        for(int i=0;i<wires.length;i++){
            graph[wires[i][0]].add(wires[i][1]);
            graph[wires[i][1]].add(wires[i][0]);
        }
        for(int i=0; i<wires.length;i++){
            int a = wires[i][0];
            int b = wires[i][1];
            boolean[] visited = new boolean[n+1];
            graph[a].remove(Integer.valueOf(b));
            graph[b].remove(Integer.valueOf(a));
            
            int cnt = dfs(i+1, visited);
            cnt = Math.abs(cnt - (n-cnt));
            answer = Math.min(cnt, answer);
            graph[a].add(b);
            graph[b].add(a);
        }
        
        
        return answer;
    }
    public static int dfs(int start, boolean[] visited){
        visited[start] = true;
        int cnt = 1;
        
        for(int next: graph[start]){
            if(!visited[next]){
                cnt += dfs(next,visited);
            }
        }
        return cnt;
    }
}